### PR TITLE
Implemented cursors using Array block APIs

### PR DESCRIPTION
**Pull request by stuartcarnie**
`TSMReader` and `KeyCursor` provide new APIs to use the Array TSM block decoders. Additionally, new cursors utilize these array APIs. 

Relevant notes:

* requires PR #10024 to be merged first in order for tests to pass
* array cursors have been reimplemented to use a 2-way merge strategy, as the old process was identified as a hot spot
* storage RPC end point uses new array cursors
    * batch cursors must remain until platform is updated
* the batch cursors can be easily enabled to test correctness and compare performance
    * this is evidenced by a couple of lines of commented code which allows switching between the two implementations

### CODE REVIEW

**Comment by stuartcarnie** [comment]
@nathanielc a lot is code-generated, so it is even easier to fix if there is a problem

File: tsdb/engine/tsm1/reader.gen.go [hunk 1]
```diff
@@ -0,0 +1,285 @@
+// Generated by tmpl
+// https://github.com/benbjohnson/tmpl
```
[the remark this answers is not in the document]
Comment (stuartcarnie) [reply tsdb/engine/tsm1/reader.gen.go]: Yes – only a single template is using the new `_tools/tmpl` for now

**Comment by hercules-influx** [comment]
During a run of [megacheck](https://github.com/dominikh/go-tools) the following issues were discovered:
```/tmp/391863931/src/github.com/influxdata/influxdb/cmd/store/query/query.go:390:5: this value of line is never used (SA4006)
/tmp/391863931/src/github.com/influxdata/influxdb/cmd/store/query/query.go:404:5: this value of line is never used (SA4006)
/tmp/391863931/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/array_cursor.gen.go:116:6: this value of tvals is never used (SA4006)
/tmp/391863931/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/array_cursor.gen.go:396:5: this value of tvals is never used (SA4006)
/tmp/391863931/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/array_cursor.gen.go:673:6: this value of tvals is never used (SA4006)
/tmp/391863931/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/array_cursor.gen.go:954:6: this value of tvals is never used (SA4006)
/tmp/391863931/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/array_cursor.gen.go:1235:6: this value of tvals is never used (SA4006)
/tmp/391863931/src/github.com/influxdata/influxdb/services/storage/batch_cursor.go:18:6: func newAggregateBatchCursor is unused (U1000)
/tmp/391863931/src/github.com/influxdata/influxdb/services/storage/batch_cursor.go:34:6: func newSumBatchCursor is unused (U1000)
/tmp/391863931/src/github.com/influxdata/influxdb/services/storage/batch_cursor.go:48:6: func newCountBatchCursor is unused (U1000)
/tmp/391863931/src/github.com/influxdata/influxdb/services/storage/batch_cursor.go:74:6: type multiShardBatchCursors is unused (U1000)
/tmp/391863931/src/github.com/influxdata/influxdb/services/storage/batch_cursor.go:88:6: func newMultiShardBatchCursors is unused (U1000)
/tmp/391863931/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/cursor_iterator.go:14:6: type cursorIterator is unused (U1000)
```

**Comment by hercules-influx** [comment]
During a run of [megacheck](https://github.com/dominikh/go-tools) the following issues were discovered:
```/tmp/081982053/src/github.com/influxdata/influxdb/cmd/store/query/query.go:390:5: this value of line is never used (SA4006)
/tmp/081982053/src/github.com/influxdata/influxdb/cmd/store/query/query.go:404:5: this value of line is never used (SA4006)
/tmp/081982053/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/array_cursor.gen.go:116:6: this value of tvals is never used (SA4006)
/tmp/081982053/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/array_cursor.gen.go:396:5: this value of tvals is never used (SA4006)
/tmp/081982053/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/array_cursor.gen.go:673:6: this value of tvals is never used (SA4006)
/tmp/081982053/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/array_cursor.gen.go:954:6: this value of tvals is never used (SA4006)
/tmp/081982053/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/array_cursor.gen.go:1235:6: this value of tvals is never used (SA4006)
/tmp/081982053/src/github.com/influxdata/influxdb/services/storage/batch_cursor.go:18:6: func newAggregateBatchCursor is unused (U1000)
/tmp/081982053/src/github.com/influxdata/influxdb/services/storage/batch_cursor.go:34:6: func newSumBatchCursor is unused (U1000)
/tmp/081982053/src/github.com/influxdata/influxdb/services/storage/batch_cursor.go:48:6: func newCountBatchCursor is unused (U1000)
/tmp/081982053/src/github.com/influxdata/influxdb/services/storage/batch_cursor.go:74:6: type multiShardBatchCursors is unused (U1000)
/tmp/081982053/src/github.com/influxdata/influxdb/services/storage/batch_cursor.go:88:6: func newMultiShardBatchCursors is unused (U1000)
/tmp/081982053/src/github.com/influxdata/influxdb/tsdb/engine/tsm1/cursor_iterator.go:14:6: type cursorIterator is unused (U1000)
```

**Comment by hercules-influx** [comment]
During a run of [megacheck](https://github.com/dominikh/go-tools) the following issues were discovered:
```/tmp/394463711/src/github.com/influxdata/influxdb/cmd/store/query/query.go:390:5: this value of line is never used (SA4006)
/tmp/394463711/src/github.com/influxdata/influxdb/cmd/store/query/query.go:404:5: this value of line is never used (SA4006)
```

**Comment by hercules-influx** [comment]
During a run of [megacheck](https://github.com/dominikh/go-tools) the following issues were discovered:
```/tmp/072507753/src/github.com/influxdata/influxdb/cmd/store/query/query.go:390:5: this value of line is never used (SA4006)
```

**Comment by hercules-influx** [comment]
During a run of [megacheck](https://github.com/dominikh/go-tools) the following issues were discovered:
```/tmp/466819943/src/github.com/influxdata/influxdb/cmd/store/query/query.go:390:5: this value of line is never used (SA4006)
```

**Pull request merged**